rfcs/999-kustomize.adoc

```diff
@@ -0,0 +1,197 @@
+      regions:
+      - name: eu-nl
+      capabilities: []
+    kustomization:
+      image:
+        name: juliogreff/bikerental
+        tag: deadcafe
+      bases:
+        - https://example.com/core
+      namePrefix: bikerental-
+      commonLabels:
+        app: bikerental
+      patchesStrategicMerge:
+        - https://example.com/sidecars/sidecar.yaml
+    strategy:
+      steps:
+      - name: full on
+        capacity:
+          incumbent: 0
+          contender: 100
+        traffic:
+          incumbent: 0
+          contender: 100
+----
+
+Note the URL `.spec.template.kustomization` points to. That's the application's
+configuration bundle. As the `kustomize` CLI tool, Shipper will also look for a
+`kustomization.yaml` inside that path.
+
+.https://example.com/sidecars/sidecar.yaml
+[source,yaml]
+----
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: app # <1>
+spec:
+  template:
+    spec:
+      containers:
+        - name: sidecar
+          image: sidecar
+----
+<1> There's a very important downside to be noted, in which sidecar patches
+must know the name of the deployment they're patching (minus any prefixes and
+suffixes), so proper conventions will be essential.
+
+.https://example.com/core/kustomization.yaml
+[source,yaml]
+----
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+  - deployment.yaml
+  - service.yaml   # <1>
+  - configmap.yaml # <1>
+----
+
+.https://example.com/core/deployment.yaml
+[source,yaml]
+----
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: app
+spec:
+  replicas: 1
+  template:
+    spec:
+      containers:
+      - name: app
+        image: app
+        ports:
+          - containerPort: 80
+----
+<1> Kustomizations can include as many of any Kubernetes resources as they
+need. As those are not of consequence to this RFC, they are omitted for
+brevity.
+
+The kustomizations above cover the application level configuration. At release
+time, Shipper still needs to apply its own little bit of configuration with
+data from the Release object it just created. That kustomization would be the
+equivalent of the following YAML snippet. It injects references to the release
+hash through `nameSuffix` and `commonLabels`, adjusts the amount of replicas
+and points to the image defined by the application object, all much in the same
+way we currently do with Helm charts.
+
+.Kustomization applied by Shipper for every release
+[source,yaml]
+----
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+bases:
+  - https://example.com/juliogreff/bikerental/kustomization.yaml
+nameSuffix: -deadbeef
+commonLabels:
+  release: bikerental-deadbeef
+replicas:
+  - name: app
+    count: 12
+images:
+  - name: app
+    newName: juliogreff/bikerental
+    newTag: deadcafe
+----
+
+== Open questions
+
+* The migration path for Shipper to move from Helm charts to Kustomize is
+  entirely unclear. We'll probably need to have both ways of packaging
```

Comment:
Yeah, I think dual life is inevitable, at least for a while. It _might_ be possible to automatically convert 'values' sections to patches, but it might not be worth the effort. My first cut would probably be just distinct 'chart' and 'kustomize' fields, a validation in the webhook which prevents both from being populated at once, or 'values' from being populated if 'kustomize' is.